libraries/Microsoft.Bot.Builder.AI.Orchestrator/OrchestratorAdaptiveRecognizer.cs

```diff
@@ -0,0 +1,334 @@
+#pragma warning restore CA2227 // Collection properties should be read only
+
+        /// <summary>
+        /// Gets or sets the disambiguation score threshold.
+        /// </summary>
+        /// <value>
+        /// Recognizer returns ChooseIntent (disambiguation) if other intents are classified within this score of the top scoring intent.
+        /// </value>
+        [JsonProperty("disambiguationScoreThreshold")]
+        public NumberExpression DisambiguationScoreThreshold { get; set; } = 0.05F;
+
+        /// <summary>
+        /// Gets or sets detect ambiguous intents.
+        /// </summary>
+        /// <value>
+        /// When true, recognizer will look for ambiguous intents (intents with close recognition scores from top scoring intent).
+        /// </value>
+        [JsonProperty("detectAmbiguousIntents")]
+        public BoolExpression DetectAmbiguousIntents { get; set; } = false;
+
+        /// <summary>
+        /// Return recognition results.
+        /// </summary>
+        /// <param name="dc">Context object containing information for a single turn of conversation with a user.</param>
+        /// <param name="activity">The incoming activity received from the user. The Text property value is used as the query text for QnA Maker.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <param name="telemetryProperties">Additional properties to be logged to telemetry with the LuisResult event.</param>
+        /// <param name="telemetryMetrics">Additional metrics to be logged to telemetry with the LuisResult event.</param>
+        /// <returns>A <see cref="RecognizerResult"/> containing the QnA Maker result.</returns>
+        public override async Task<RecognizerResult> RecognizeAsync(DialogContext dc, Schema.Activity activity, CancellationToken cancellationToken, Dictionary<string, string> telemetryProperties = null, Dictionary<string, double> telemetryMetrics = null)
+        {
+            var text = activity.Text ?? string.Empty;
+            var detectAmbiguity = DetectAmbiguousIntents.GetValue(dc.State);
+
+            _modelPath = ModelPath.GetValue(dc.State);
+            _snapshotPath = SnapshotPath.GetValue(dc.State);
+
+            InitializeModel();
+
+            var recognizerResult = new RecognizerResult()
+            {
+                Text = text,
+                Intents = new Dictionary<string, IntentScore>(),
+            };
+
+            if (string.IsNullOrWhiteSpace(text))
+            {
+                // nothing to recognize, return empty recognizerResult
+                return recognizerResult;
+            }
+
+            if (EntityRecognizers.Count != 0)
+            {
+                // Run entity recognition
+                recognizerResult = await RecognizeEntitiesAsync(dc, activity, recognizerResult).ConfigureAwait(false);
```

Comment:
Here there is no justification for passing in the recognizerResult _and_ return it. I would pick one or the other, and generally it's better to return things when you can - though that's more of a style thing perhaps.

Anyhow it reads weird when you are passing in and returning in this scenario (there are times, for fluent apis perhaps) when we do that but not here.

Reply:
Fixed.